File: frontend/src/services/api/orderService.ts
/**
 * Updated Order Service with ApiResponse support
 * Includes pagination support for order lists
 */

import apiClient from './apiClient';
import { OrderResponse, OrderPreviewResponse, OrderRequest } from '../../types/order';
import { PaginationInfo, ApiResponse } from '../../types/apiResponse';

const ORDER_API = '/orders';

export interface PaginatedOrdersResponse {
  items: OrderResponse[];
  pagination: PaginationInfo;
}

export const orderService = {
  /**
   * Create a new order
   * Response: ApiResponse<OrderResponse>
   */
  async createOrder(userId: string, orderRequest: OrderRequest): Promise<OrderResponse> {
    const response = await apiClient.post<ApiResponse<OrderResponse>>(
      `${ORDER_API}/${userId}`,
      orderRequest
    );
    return response.data.data;
  },

  /**
   * Preview order before checkout
   * Response: ApiResponse<OrderPreviewResponse>
   */
  async previewOrder(orderRequest: OrderRequest): Promise<OrderPreviewResponse> {
    const response = await apiClient.post<ApiResponse<OrderPreviewResponse>>(
      `${ORDER_API}/preview`,
      orderRequest
    );
    return response.data.data;
  },

  /**
   * Get all orders (ADMIN only, non-paginated)
   * Response: ApiResponse<OrderResponse[]>
   */
  async getAllOrders(): Promise<OrderResponse[]> {
    const response = await apiClient.get<ApiResponse<OrderResponse[]>>(`${ORDER_API}/all`);
    return response.data.data;
  },

  /**
   * Get all orders with pagination (ADMIN only)
   * Response: ApiResponse<OrderResponse[]> with pagination info
   * 
   * Example:
   * const { data: orders, pagination } = await orderService.getAllOrdersPaginated(0, 10, 'createdAt');
   */
  async getAllOrdersPaginated(
    page: number = 0,
    size: number = 10,
    sort: string = 'createdAt'
  ): Promise<PaginatedOrdersResponse> {
    const response = await apiClient.get<ApiResponse<OrderResponse[]>>(`${ORDER_API}`, {
      params: { page, size, sort },
    });
    
    return {
      items: response.data.data,
      pagination: response.data.pagination!,
    };
  },

  /**
   * Get order by ID
   * Response: ApiResponse<OrderResponse>
   */
  async getOrderById(orderId: string): Promise<OrderResponse> {
    const response = await apiClient.get<ApiResponse<OrderResponse>>(`${ORDER_API}/${orderId}`);
    return response.data.data;
  },

  /**
   * Get user orders (non-paginated)
   * Response: ApiResponse<OrderResponse[]>
   */
  async getUserOrders(userId: string): Promise<OrderResponse[]> {
    const response = await apiClient.get<ApiResponse<OrderResponse[]>>(
      `${ORDER_API}/user/${userId}/all`
    );
    return response.data.data;
  },

  /**
   * Get user orders with pagination
   * Response: ApiResponse<OrderResponse[]> with pagination info
   * 
   * Example:
   * const { data: orders, pagination } = await orderService.getUserOrdersPaginated(userId, 0, 10);
   */
  async getUserOrdersPaginated(
    userId: string,
    page: number = 0,
    size: number = 10,
    sort: string = 'createdAt'
  ): Promise<PaginatedOrdersResponse> {
    const response = await apiClient.get<ApiResponse<OrderResponse[]>>(
      `${ORDER_API}/user/${userId}`,
      {
        params: { page, size, sort },
      }
    );
    
    return {
      items: response.data.data,
      pagination: response.data.pagination!,
    };
  },

  /**
   * Cancel order
   * Response: ApiResponse<OrderResponse>
   */
  async cancelOrder(orderId: string): Promise<OrderResponse> {
    const response = await apiClient.delete<ApiResponse<OrderResponse>>(
      `${ORDER_API}/${orderId}`
    );
    return response.data.data;
  },

  /**
   * Update order status (ADMIN only)
   * Response: ApiResponse<OrderResponse>
   * 
   * Valid status values: PENDING, CONFIRMED, PROCESSING, SHIPPED, DELIVERED, CANCELLED
   */
  async updateOrderStatus(orderId: string, status: string): Promise<OrderResponse> {
    const response = await apiClient.patch<ApiResponse<OrderResponse>>(
      `${ORDER_API}/${orderId}/${status}`
    );
    return response.data.data;
  },
};

File: backend/src/main/java/com/nexoracommerce/order/service/OrderService.java
package com.nexoracommerce.order.service;

import com.nexoracommerce.common.enums.OrderStatus;
import com.nexoracommerce.order.dto.request.OrderRequest;
import com.nexoracommerce.order.dto.request.OrderStatusChangeRequest;
import com.nexoracommerce.order.dto.response.OrderResponse;
import com.nexoracommerce.order.dto.response.OrderPreviewResponse;
import com.nexoracommerce.order.dto.response.OrderStatusHistoryResponse;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

/**
 * Service interface for Order operations
 */
public interface IOrderService {

    /**
     * Create a new order
     * @param request OrderRequest containing order details and items
     * @param userId the user creating the order
     * @return created OrderResponse
     */
    OrderResponse createOrder(OrderRequest request, String userId);

    /**
     * Get order by ID
     * @param orderId the order ID
     * @return OrderResponse with full details
     */
    OrderResponse getOrderById(String orderId);

    /**
     * Get all orders for a user
     * @param userId the user ID
     * @return list of OrderResponse
     */
    List<OrderResponse> getUserOrders(String userId);

    /**
     * Get all orders for a user with pagination
     * @param userId the user ID
     * @param pageable pagination parameters
     * @return Page of OrderResponse
     */
    Page<OrderResponse> getUserOrdersPageable(String userId, Pageable pageable);

    /**
     * Cancel an order
     * @param orderId the order ID
     * @return updated OrderResponse
     */
    OrderResponse cancelOrder(String orderId);

    /**
     * Update order status with audit trail
     * @param request containing orderId, newStatus, and optional reason
     * @return updated OrderResponse
     */
    OrderResponse updateOrderStatus(OrderStatusChangeRequest request);

    /**
     * Get order status history
     * @param orderId the order ID
     * @return list of OrderStatusHistoryResponse
     */
    List<OrderStatusHistoryResponse> getOrderStatusHistory(String orderId);

    /**
     * Preview order price before checkout
     * @param request OrderRequest containing order details
     * @return OrderPreviewResponse with price breakdown
     */
    OrderPreviewResponse previewOrder(OrderRequest request);

    /**
     * Get all orders (admin only)
     * @return list of all OrderResponse
     */
    List<OrderResponse> getAllOrders();

    /**
     * Get all orders with pagination (admin only)
     * @param pageable pagination parameters
     * @return Page of OrderResponse
     */
    Page<OrderResponse> getAllOrdersPageable(Pageable pageable);

    /**
     * Get orders by status with pagination
     * @param status the order status
     * @param pageable pagination parameters
     * @return Page of OrderResponse
     */
    Page<OrderResponse> getOrdersByStatus(OrderStatus status, Pageable pageable);

    /**
     * Get unpaid orders for a user
     * @param userId the user ID
     * @param pageable pagination parameters
     * @return Page of OrderResponse
     */
    Page<OrderResponse> getUnpaidOrdersByUser(String userId, Pageable pageable);
}
